### DISCUSSION TITLE
Multiple containers, update content based on routing

### DISCUSSION THREAD

**Discussion by jacoblindgren** [swup/swup#546] [Help] [answered]
I have a page that is split down the middle, and swup's `containers` option is set to `["#left-content", "#right-content"]`. I'm curious what the best way is, or if it's possible, to tell swup to update one of the two containers based on the route the user visits.

For example when updating `#right-content` I mostly want `#left-content` to remain untouched, and vice versa when updating `#left-content` I'd like `#right-content` to remain in place.

**Comment by hirasso** [comment, marked as answer]
Update: [swup 4](https://swup.js.org/announcements/swup-4/) allows [customizing containers](https://swup.js.org/visit/#dynamic-containers) for each visit on the fly:

```js
swup.hooks.on('visit:start', (visit) => {
  visit.containers = ['#left-content'];
})
```

Previous answer for swup 3:

> Tying containers to route fragments is not possible with swup out of the box right now. You could probably work out a solution by deciding on the server what do return for each container based on the current URL. Maybe [Morph Plugin](https://github.com/daun/swup-morph-plugin) could also come in handy for your use case.

**Comment by daun** [comment]
You could try listening to the [clickLink event](https://swup.js.org/events#list-of-all-events) and modifying swup's `containers` option on the fly.

**Comment by hirasso** [comment]
@daun Using the `clickLink` event also sounds like a nice idea. It would just have to be combined with the `popState` event, to support the history (back/forward) buttons, as well. @jacoblindgren Maybe this could be a starting point:

```js
const swup = new Swup({
  containers: ['#left-content', '#right-content']
})
const adjustContainers = (url) => {
  if (/* url matches condition */ true) {
    swup.options.containers = ['#left-content'];
    return;
  }
  if (/* url matches other condition */ true) {
    swup.options.containers = ['#right-content'];
    return;
  }
  // no condition matched
  swup.options.containers = ['#left-content', '#right-content'];
};
swup.on("clickLink", (event) => adjustContainers(event.delegateTarget.href));
swup.on("popState", () => adjustContainers(window.location.href));

```

**Reply by jacoblindgren** [reply]
This is incredibly helpful, thank you. I'm experimenting with it now and I've built each `if` statement so that it addresses whether a given link should be replacing content from the `#left-content` or `#right-content`, but it seems like there are still some wires getting crossed at some point in the process, as a click which targets `#right-content` is apparently grabbing the container but putting it in the place of `#left-content`. (Seen here: https://streamable.com/t1lqrj, which is running a click tied to the `if` statement which sets `swup.options.containers = ['#right-content']`)

I'm doing `console.log(swup.options.containers)` before the `return` and the containers are correctly being changed, but it's like there's some ghost of containers past that is still mixing things up?

I realize this is probably incredibly difficult, near impossible, to imagine without an example, but I really appreciate your time regardless.

**Reply by daun** [reply]
@jacoblindgren Swup expects the number of containers to be the same for every page load, so it's probably getting confused about two becoming one and vice-versa. Try adding dummy selectors in place of the ignored ones and let us know if it works as expected:

```js
swup.options.containers = ['#left-content', '#non-existing-div'] // for left-hand only
swup.options.containers = ['#non-existing-div', '#right-content'] // for right-hand only
```

**Reply by jacoblindgren** [reply]
Great idea. Same behavior, sadly, but it seems like the answer (or the "wires crossing") lies somewhere in here. I tried the arrangement you suggested, and also naively tried:

```javascript
swup.options.containers = ['#left-content', '#non-existing-div'] // for left-hand only
swup.options.containers = ['#right-content', '#non-existing-div'] // for right-hand only
```

in case it was tied to the order (`[target, ignore]`) instead of just the `#` itself. It's returning a soft error that it cannot find the `#non-existing-div`, of course, but otherwise behaving the same.

What's interesting is that the behavior recorded in that video above also works the other way around: `#left-content` is simply replaced by `#right-content` for clicks targeting the right, but then when clicking back on "left" links `#left-content` then gets reverted to being `#left-content`. So, Swup really fixates on the first container.

**Reply by daun** [reply]
What if it's not a non-existent div but a hidden one?

```js
swup.options.containers = ['#left-content', '#visually-hidden-div']
swup.options.containers = ['#visually-hidden-div', '#right-content']
```

**Reply by jacoblindgren** [reply]
This makes for really strange but potentially revealing behavior, although it's hard to describe succinctly:

- on the first "left" click, `#right-content` is replaced with `#visually-hidden-div`, but `#left-content` is successfully swapped (I guess Swup remembers there is a second entry in the containers array, but now thinks `#right-container` is the "destination" for `#visually-hidden-div`)
- on a second "left" click, I get a `[swup] Mismatched number of containers found on new page.` error and `Uncaught TypeError: Cannot read properties of undefined (reading 'setAttribute')` and it breaks
- "right" clicks sometimes work, after a page refresh, but `#left-content` is being replaced by `#visually-hidden-div`

I can't put my finger on the inner workings, but it seems like "once a container, always a container" is the issue? Seems like clicking on a "right" link (`swup.options.containers = ['#visually-hidden-div', '#right-content']`) isn't discarding the "swapping" event, or "disposing" of it inside the hidden container, and instead is telling `#visually-hidden-div`(the new value) to replace `#left-content` (the original value).

I don't know why anything past the first "left" click is failing/causing all those errors, though.

Greatly appreciative of all the thought and time given to this, thanks again. If you have any other directions to point in or reckon it's hit a wall both are incredibly useful!

**Reply by daun** [reply]
@jacoblindgren I wonder if you could create a simplified test case by [forking this template](https://codesandbox.io/s/swup-test-case-template-6ci5si)? I'd love to fiddle with this but don't have time for setting up a test environment currently.

**Reply by jacoblindgren** [reply]
Here is a [forked version](https://codesandbox.io/s/swup-multiple-b6emb7?file=/index.js:477-494) representative of my use case. Hopefully it's simplified enough. The `url.substr(0, 31)` in the if statement is awkward but checks if it's an `/articles` link being clicked on or not, or in other words whether it's a `page-*.html` file (which should show in `#left-content`) or whether it's an `article-*.html` file, which should leave `#left-content` untouched and only alter `#right-content`.

**Reply by daun** [reply]
I gave it a try. See this [CodeSandbox with a working version](https://codesandbox.io/s/swup-multiple-forked-lxnes3). This solution uses the [morph plugin](https://github.com/daun/swup-morph-plugin) for updating the containers, leaving swup with a single hidden empty dummy container.

**Reply by jacoblindgren** [reply]
This is incredible, thank you so much (and @hirasso too)! I really appreciate your help and time.
